### PR TITLE
Add challenges and README for session 04 - SQL injection

### PR DESCRIPTION
**Pull request by gabrielmocanu**
Add challenges and README for session 04 - SQL injection

### CODE REVIEW

File: sql-injection/README.md [hunk 1]
```diff
@@ -176,25 +211,29 @@ When you have determined the number of columns returned by the original query an
...
+Without this information, you would be left trying to guess the names of tables and columns.
+In fact, all modern databases provide ways of examining the database structure, to determine what tables and columns it contains.
 
 ### Retrieving multiple values within a single column
 
 In the preceding example, suppose instead that the query only returns a single column.
 
-You can easily retrieve multiple values together within this single column by concatenating the values together, ideally including a suitable separator to let you distinguish the combined values. For example, on Oracle you could submit the input:
+You can easily retrieve multiple values together within this single column by concatenating the values together, ideally including a suitable separator to let you distinguish the combined values.
+For example, on Oracle you could submit the input:
```
Comment (carrrina) [inline sql-injection/README.md:231]:
```suggestion
For example, in Oracle you could submit the input:
```

File: sql-injection/README.md [hunk 2]
```diff
@@ -176,25 +211,29 @@ When you have determined the number of columns returned by the original query an
...
 
 `' UNION SELECT username || '~' || password FROM users--`
 
-This uses the double-pipe sequence `||` which is a string concatenation operator on Oracle. The injected query concatenates together the values of the `username` and `password` fields, separated by the `~` character.
+This uses the double-pipe sequence `||` which is a string concatenation operator on Oracle.
```
Comment (carrrina) [inline sql-injection/README.md:235]:
```suggestion
This uses the double-pipe sequence `||` which is a string concatenation operator in Oracle.
```

File: sql-injection/README.md [hunk 3]
```diff
@@ -206,31 +245,37 @@ carlos~montoya
...
 ## Examining the database
 
-Following the initial identification of an SQL injection vulnerability, it is generally useful to obtain some information about the database itself. This information can often pave the way for further exploitation.
+Following the initial identification of an SQL injection vulnerability, it is generally useful to obtain some information about the database itself.
+This information can often pave the way for further exploitation.
 
-You can query the version details for the database. The way that this is done depends on the database type, so you can infer the database type from whichever technique works. For example, on Oracle you can execute:
+You can query the version details for the database.
+The way that this is done depends on the database type, so you can infer the database type from whichever technique works.
+For example, on Oracle you can execute:
```
Comment (carrrina) [inline sql-injection/README.md:258]:
```suggestion
For example, in Oracle you can execute:
```

File: sql-injection/README.md [hunk 4]
```diff
@@ -295,47 +349,61 @@ We can continue this process to systematically determine the full password for t
 
 **NOTE:**
 
-The `SUBSTRING` function is called `SUBSTR` on some types of database. For more details, see the [SQL injection cheat sheet](https://portswigger.net/web-security/sql-injection/cheat-sheet).
+The `SUBSTRING` function is called `SUBSTR` on some types of database.
```
Comment (carrrina) [inline sql-injection/README.md:352]:
```suggestion
The `SUBSTRING` function is called `SUBSTR` on some types of databases.
```

Comment (carrrina) [inline sql-injection/README.md:352]:
```suggestion
The `SUBSTRING` function is called `SUBSTR` on some types of databases.
```

File: sql-injection/README.md [hunk 5]
```diff
@@ -295,47 +349,61 @@ We can continue this process to systematically determine the full password for t
...
 
 Using this technique, we can retrieve data in the way already described, by systematically testing one character at a time:
 
 `'; IF (SELECT COUNT(username) FROM Users WHERE username = 'Administrator' AND SUBSTRING(password, 1, 1) > 'm') = 1 WAITFOR DELAY '0:0:{delay}'--`
 
 ### Second Order SQL Injection
+
 A Second Order Injection is the same as a traditional injection attack but the payload is already stored in the database intentionally placed so that it can be triggered in another area of code.
 
 Let’s look at an example and see how easy it to exploit this vulnerability.
```
Comment (carrrina) [inline sql-injection/README.md:404]:
```suggestion
Let’s look at an example and see how easy it is to exploit this vulnerability.
```

Comment (carrrina) [inline sql-injection/README.md:404]:
```suggestion
Let’s look at an example and see how easy it is to exploit this vulnerability.
```

File: sql-injection/README.md [hunk 6]
```diff
@@ -388,26 +457,36 @@ statement.setString(1, input);
...
+With this tool you can see all the requests that are executed when interacting with a web application.
+You can change the requests to see different behaviors and you can set automatic requests with different parameters.
 
 ## SQLmap
-[SQLmap](https://sqlmap.org/) is an open source penetration testing tool that automates the process of detecting and exploiting SQL injection flaws and taking over of database servers. It comes with a powerful detection engine, many niche features for the ultimate penetration tester and a broad range of switches lasting from database fingerprinting, over data fetching from the database, to accessing the underlying file system and executing commands on the operating system via out-of-band connections. You can see some of the features in this recording.
-[![asciicast](https://asciinema.cs.pub.ro/a/N1r0DeFboLW4GXj0rZb4jsYeU.png)](https://asciinema.cs.pub.ro/a/N1r0DeFboLW4GXj0rZb4jsYeU)
+
+[SQLmap](https://sqlmap.org/) is an open source penetration testing tool that automates the process of detecting and exploiting SQL injection flaws and taking over of database servers.
```
Comment (carrrina) [inline sql-injection/README.md:478]:
```suggestion
[SQLmap](https://sqlmap.org/) is an open source penetration testing tool that automates the process of detecting and exploiting SQL injection flaws and taking over database servers.
```

Comment (carrrina) [inline sql-injection/README.md:478]:
```suggestion
[SQLmap](https://sqlmap.org/) is an open source penetration testing tool that automates the process of detecting and exploiting SQL injection flaws and taking over database servers.
```

File: sql-injection/README.md [hunk 7]
```diff
@@ -143,21 +167,32 @@ If the number of nulls does not match the number of columns, the database return
...
+- The reason for using `NULL` as the values returned from the injected `SELECT` query is that the data types in each column must be compatible between the original and the injected queries.
+  Since `NULL` is convertible to every commonly used data type, using `NULL` maximizes the chance that the payload will succeed when the column count is correct.
+- On Oracle, every `SELECT` query must use the FROM keyword and specify a valid table.
+  There is a built-in table on Oracle called `DUAL` which can be used for this purpose.
+  So the injected queries on Oracle would need to look like: `' UNION SELECT NULL FROM DUAL--`.
+- The payloads described use the double-dash comment sequence `--` to comment out the remainder of the original query following the injection point.
+  On MySQL, the double-dash sequence must be followed by a space. Alternatively, the hash character `#` can be used to identify a comment.
```
Comment (carrrina) [inline sql-injection/README.md:185]:
```suggestion
  In MySQL, the double-dash sequence must be followed by a space. Alternatively, the hash character `#` can be used to identify a comment.
```

Comment (carrrina) [inline sql-injection/README.md:185]:
```suggestion
  In MySQL, the double-dash sequence must be followed by a space. Alternatively, the hash character `#` can be used to identify a comment.
```

**Comment by carrrina** [comment]
I think you made a mistake - for the challenges, only the `sol` folder (and maybe the `README.md` file) should be present in this public repo

**Comment by carrrina** [comment]
I think you made a mistake - for the challenges, only the `sol` folder (and maybe the `README.md` file) should be present in this public repo

**Review by carrrina — CHANGES_REQUESTED** [review]

**Comment by carrrina** [comment]
All the `solution` scripts should be named `solution.sh` as per [here](https://github.com/security-summer-school/template-internal)

**Comment by gabrielmocanu** [comment]
@carrrina 

Can you please look again at this PR?
I resolve all the issues you mentioned. Thanks!

File: sql-injection/README.md [hunk 8]
```diff
@@ -388,40 +457,58 @@ statement.setString(1, input);
...
-**12.** [Out-of-band Application Security Testing](https://portswigger.net/web-security/sql-injection/blind/lab-out-of-band)  
+# Activities
+
+**1.** [Demo SQL Injection](https://sss-ctf.security.cs.pub.ro/challenges)
+**2.** [Blacklist](https://sss-ctf.security.cs.pub.ro/challenges)
+**3.** [Nightmare Store](https://sss-ctf.security.cs.pub.ro/challenges)
+**4.** [One by one](https://sss-ctf.security.cs.pub.ro/challenges)
+**5.** [T0p S3cr3t](https://sss-ctf.security.cs.pub.ro/challenges)
+
+# Other Activities
```
Comment (carrrina) [inline sql-injection/README.md:501]: I think this line is useless, they are all activities

**Review by carrrina — CHANGES_REQUESTED** [review]

**Comment by carrrina** [comment]
You can delete the old session and modify the repo README to point at this one

**Comment by gabrielmocanu** [comment]
@carrrina 

Can you take a look again?

**Review by carrrina — APPROVED** [review]

**Pull request merged**